include/nbl/builtin/hlsl/bxdf/common.hlsl

```diff
@@ -48,358 +139,976 @@ struct Basic
-  }
-
-  float3 getTangentSpaceV() {return float3(Tdot,BdotV,Isotropic<RayDirInfo>::NdotV);}
-  // WARNING: its the transpose of the old GLSL function return value!
-  float3x3 getTangentFrame() {return float3x3(T,B,Isotropic<RayDirInfo>::N);}
-
-  float3 T;
-  float3 B;
-  float3 TdotV;
-  float3 BdotV;
+    using ray_dir_info_type = RayDirInfo;
+    using scalar_type = typename RayDirInfo::scalar_type;
+    using vector3_type = typename RayDirInfo::vector3_type;
+    using matrix3x3_type = matrix<scalar_type, 3, 3>;
+    using isotropic_type = SIsotropic<RayDirInfo>;
+
+    // WARNING: Changed since GLSL, now arguments need to be normalized!
+    static SAnisotropic<RayDirInfo> create(
+        NBL_CONST_REF_ARG(isotropic_type) isotropic,
+        NBL_CONST_REF_ARG(vector3_type) normalizedT,
+        NBL_CONST_REF_ARG(vector3_type) normalizedB
+    )
+    {
+        SAnisotropic<RayDirInfo> retval;
+        retval.isotropic = isotropic;
+
+        retval.T = normalizedT;
+        retval.B = normalizedB;
+
+        retval.TdotV = nbl::hlsl::dot<vector3_type>(retval.isotropic.getV().getDirection(), retval.T);
+        retval.BdotV = nbl::hlsl::dot<vector3_type>(retval.isotropic.getV().getDirection(), retval.B);
+
+        return retval;
+    }
+    static SAnisotropic<RayDirInfo> create(NBL_CONST_REF_ARG(isotropic_type) isotropic, NBL_CONST_REF_ARG(vector3_type) normalizedT)
+    {
+        return create(isotropic, normalizedT, cross(isotropic.getN(), normalizedT));
+    }
+    static SAnisotropic<RayDirInfo> create(NBL_CONST_REF_ARG(isotropic_type) isotropic)
+    {
+        vector3_type T, B;
+        math::frisvad<vector3_type>(isotropic.getN(), T, B);
+        return create(isotropic, nbl::hlsl::normalize<vector3_type>(T), nbl::hlsl::normalize<vector3_type>(B));
+    }
+
+    vector3_type getT() NBL_CONST_MEMBER_FUNC { return T; }
+    vector3_type getB() NBL_CONST_MEMBER_FUNC { return B; }
+    scalar_type getTdotV() NBL_CONST_MEMBER_FUNC { return TdotV; }
+    scalar_type getBdotV() NBL_CONST_MEMBER_FUNC { return BdotV; }
+
+    vector3_type getTangentSpaceV() NBL_CONST_MEMBER_FUNC { return vector3_type(TdotV, BdotV, isotropic.NdotV); }
+    matrix3x3_type getToTangentSpace() NBL_CONST_MEMBER_FUNC { return matrix3x3_type(T, B, isotropic.N); }
+    matrix3x3_type getFromTangentSpace() NBL_CONST_MEMBER_FUNC { return nbl::hlsl::transpose<matrix3x3_type>(matrix3x3_type(T, B, isotropic.N)); }
+
+    isotropic_type isotropic;
+    vector3_type T;
+    vector3_type B;
+    scalar_type TdotV;
+    scalar_type BdotV;
 };
 
 }
 
+template<typename InteractionT>
+struct interaction_traits;
+
+template<typename RayDirInfo>
+struct interaction_traits<surface_interactions::SIsotropic<RayDirInfo> >
+{
+    NBL_CONSTEXPR_STATIC_INLINE bool is_aniso = false;
+};
 
-template<class RayDirInfo>
-struct LightSample
+template<typename RayDirInfo>
+struct interaction_traits<surface_interactions::SAnisotropic<RayDirInfo> >
```

Comment:
what do you need this trait for? wherever you check `interaction_traits<Int>::is_aniso` you could check `Anisotropic<Int>` instead